### DISCUSSION TITLE
wezterm.exec_domain destroys true-color support for helix

### DISCUSSION THREAD

**Discussion by favetelinguis** [wezterm/wezterm#5546] [Q&A] [answered]
### What Operating System(s) are you running on?

Linux Wayland

### Which Wayland compositor or X11 Window manager(s) are you using?

Gnome

### WezTerm version

wezterm 20240203.110809.5046fc22

### Ask your question!

I have modified the example for using exec_domain to work with distrobox instead and it work very nice except for one issue. Helix reports that it cant use true colors when started using the new domains. Below is my .wezterm.conf
```
local wezterm = require 'wezterm'
local mux = wezterm.mux
local act = wezterm.action

wezterm.on('gui-startup', function()
 local tab, pane, window = mux.spawn_window({})
 window:gui_window():maximize()
end)

local config = wezterm.config_builder()

-- Had issues where mouse cursor wher not shown
-- gsettings get org.gnome.desktop.interface cursor-size -> 32
-- gsettings get org.gnome.desktop.interface cursor-theme -> Adwaita
-- and make sure the theme is avaliaable in the distrobox container
-- /usr/share/icons/Adwaita
config.xcursor_theme = "Adwaita"
config.xcursor_size = 24

config.window_decorations = "RESIZE"


config.color_scheme = 'Catppuccin Mocha'
-- Default font is Jetbrains Mono Nerd
config.font_size = 12
config.line_height = 1.2

config.keys = {

{ key = 'l', mods = 'ALT', action = wezterm.action.ShowLauncher },  
  {
    key = 'P',
    mods = 'CTRL',
    action = wezterm.action.ActivateCommandPalette,
  },
}

-- Function to split a string by a given delimiter
function split(inputstr, sep)
    if sep == nil then
        sep = "%s"
    end
    local t = {}
    for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
        table.insert(t, str)
    end
    return t
end

-- Function to trim leading and trailing whitespace from a string
function trim(s)
    return (s:gsub("^%s*(.-)%s*$", "%1"))
end

-- Function to extract the NAME column as a table of strings
function extractNameColumn(inputstr)
    local lines = split(inputstr, "\n")
    local names = {}
    local isHeader = true
    
    for _, line in ipairs(lines) do
        -- Skip the header line
        if isHeader then
            isHeader = false
        else
            local columns = split(line, "|")
            -- Remove leading and trailing whitespace from each column
            for i = 1, #columns do
                columns[i] = trim(columns[i])
            end
            -- Add the second column (NAME) to the names table
            table.insert(names, columns[2])
        end
    end
    
    return names
end

function docker_list()
  local docker_list = {}
  local success, stdout, stderr = wezterm.run_child_process {
    "distrobox",
    "list",
  }
  for id, name in ipairs(extractNameColumn(stdout)) do
    if id and name then
      docker_list[id] = name
    end
  end
  return docker_list
end

function make_docker_label_func(name1)
  return function(name)
    return wezterm.format {
      { Foreground = { AnsiColor = 'Green' } },
      { Text = 'distrobox named ' .. name },
    }
  end
end

function make_docker_fixup_func(name)
  return function(cmd)
    cmd.args = cmd.args or { '/bin/bash' }
    local wrapped = {
      'distrobox',
      'enter',
      '-n',
      name,
      '-e',
    }
    for _, arg in ipairs(cmd.args) do
      table.insert(wrapped, arg)
    end

    cmd.args = wrapped
    return cmd
  end
end

function compute_exec_domains()
  local exec_domains = {}
  for id, name in pairs(docker_list()) do
    table.insert(
      exec_domains,
      wezterm.exec_domain(
        'distrobox:' .. name,
        make_docker_fixup_func(name),
        make_docker_label_func(name)
      )
    )
  end
  return exec_domains
end

config.exec_domains = compute_exec_domains()

return config

```

To reproduce you need distrobox installed. The do the following : 
`distrobox create wez1 -i ghcr.io/favetelinguis/polyglot:main`
`distrobox create wez2 -i ghcr.io/favetelinguis/polyglot:main`
Then open a terminal and enter
`distrobox enter wez1 -- wezterm`
This will start up wezterm, now try to open helix and you will see that true color support is avaliable.
Now inside wezterm do Alt-l and select the option to open a new tab using the domain distrobox:wez2, in the new tab open helix and you will now see that helix is unable to load any true color themes.

I would very much like to understand what is happening here but I am far from an expert. Any pointers on what I need to change to make this setup work would make me very happy.

**Comment by wez** [comment, marked as answer]
It's most likely something to do with the environment in the commands you are spawning. Compare the output from `env` between the working and non-working environments.

**Reply by favetelinguis** [reply]
You are correct, it was the environment variable COLORTERM=truecolor that for some reason gets lost. Do you happen to have some clue why this gets set when I do distrobox enter wez1 -- wezterm but not when I use use exec_domains? Do not processes started with exec_domain inherit the env?  And is there a way to have wezterm set the environment in also the spawned processes, all of the following are not set in the spawned process which maybe will cause other issues:

WEZTERM_CONFIG_DIR=/home/dsss
WEZTERM_EXECUTABLE=/usr/bin/wezterm-gui
WEZTERM_UNIX_SOCKET=/run/user/1000/wezterm/gui-sock-11509
WEZTERM_CONFIG_FILE=/home/dsss/.wezterm.lua
WEZTERM_PANE=0
WEZTERM_EXECUTABLE_DIR=/usr/bin

**Reply by wez** [reply]
processes spawned by wezterm inherit wezterm's environment, which is in turn typically inherited from the desktop environment which spawned wezterm.

When you run distrobox from your shell, it is inheriting the environment of that shell.

You may need to explicitly pass the environment down to distrobox from your `make_docker_fixup_func`; the `cmd` object should have a list of environment variables and you may need to either tell `distrobox` to inherit everything, or you may need to explicitly pass them down. I don't use distrobox, but from a quick look at its docs, it looks like the `--env` parameter can be used for that, but I'm not sure if there is a more general inherit everything flag.

To debug this sort of thing, I tend to use something like `wezterm.log_info(cmd)` to dump the cmd object so that you can see what is in there.
